### PR TITLE
Non-blocking Ring Attention

### PR DESCRIPTION
**Pull request by coreyjadams**
Update ring attention to use non blocking communication, and ensure test coverage.

This is a year or more overdue. This PR updates ring attention to enable overlap of ring attention communication and computation.

I also have test coverage to make sure the ring operations are tested in isolation, now, and not as part of operations.  And the blocking and non-blocking are compared, which helps ensure they are correct.

image

# PhysicsNeMo Pull Request

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

**Comment by coreyjadams** [comment]
I will update this PR with some scripts to run and benchmark ring attention in the `examples/minimal/ShardTensorExamples/` folder, but that will be just icing on the cake.  That might help enable testing if you want, but shouldn't block a review or anything.

File: physicsnemo/domain_parallel/shard_utils/attention_patches.py [hunk 1]
```diff
@@ -191,109 +190,97 @@ def forward(
...
         global_log_sumexp = None
 
-        # For the first iteration, use local tensors
-        current_k, current_v = k, v
-
-        # Pre-allocate a single combined buffer for k,v
-        # Find the dimension to concatenate on - should be a dim that preserves tensor structure
-        cat_dim = 0  # Usually batch or sequence dimension
+        compute_stream = torch.cuda.current_stream()
+        comm_stream = torch.cuda.Stream()
```
[the remark this answers is not in the document]
Comment (pzharrington) [reply physicsnemo/domain_parallel/shard_utils/attention_patches.py]: @coreyjadams does the commented-out/removed `comm_stream` correspond to a rabbit hole that didn't pan out trying to avoid recreating streams every fwd/bwd?

File: physicsnemo/domain_parallel/shard_utils/ring.py [hunk 2]
```diff
@@ -192,3 +217,88 @@ def perform_ring_iteration(
...
+            "perform_ring_iteration_async only supports p2p communication. "
+            f"Got: {ring_config.communication_method}"
+        )
+
+    local_group, _, _, _, id_for_send, id_for_recv = _get_ring_comm_ranks(
+        mesh, ring_config
+    )
+
+    if not tensor.is_contiguous():
+        tensor = tensor.contiguous()
```
[the remark this answers is not in the document]
Comment (pzharrington) [reply physicsnemo/domain_parallel/shard_utils/ring.py]: Docstring says the tensor `Must be contiguous.` so maybe we just raise and exception here and put contiguity-enforcement in the caller function?

File: physicsnemo/domain_parallel/shard_utils/attention_patches.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (coreyjadams) [reply physicsnemo/domain_parallel/shard_utils/attention_patches.py]: I recall complaints when I put a cuda stream outside the context, that when CPU-only users tried to use physicsnemo it broke things.  I'll update with a persistent stream that won't have to hit the driver every time we call it.

**Comment by coreyjadams** [comment]
@pzharrington I addressed your comments.  I also added a new example in the ShardTensor examples, for RingAttention specifically.

While I was updating these examples, I added the plotting scripts for both the ViT and RingAttention scripts, which enables reproduction of the ShardTensor benchmarks in the paper.  We'll need this within 2 weeks for the paper so I figured it's good to get the ball rolling.

The PR looks a lot bigger than it is because I took this opportunity, while I'm updating here, to add all the missing docstrings in ShardTensor examples that interrogate had been missing.  Sorry for adding some bloat.

File: physicsnemo/domain_parallel/shard_utils/attention_patches.py [hunk 3]
```diff
@@ -324,10 +311,18 @@ def backward(
...
+        Overlaps k/v communication with the backward attention kernel.
+        Each iteration:
+        1. Async-send k, v for the next iteration (overlaps with compute).
+        2. Wait for grad_k, grad_v from the previous iteration's send.
+        3. Compute block gradients and accumulate.
+        4. Wait for k, v receive to complete.
+        5. Async-send accumulated grad_k, grad_v (overlaps with next
+           iteration's k/v receive and compute start).
+        The final grad_k/grad_v shift uses blocking communication since
+        there is no further compute to overlap with.
```
[the remark this answers is not in the document]
Comment (coreyjadams) [reply physicsnemo/domain_parallel/shard_utils/attention_patches.py:331]: The docstring has been updated.

**Comment by coreyjadams** [comment]
Regarding the comm_stream in ring attention: There is now a factory function in `ring.py` to yield a new stream or retrieve a cached stream, based on device ordinal.  So we'll create up to one comm stream per device to be reused for all ring communications.

**Review by ktangsali — APPROVED** [review]
Not sure why I got pinged for this, but approving to unblock. Assuming that @pzharrington will be doing the primary review

**Review by pzharrington — APPROVED** [review]
Nice job. My only remaining nit is that the ViT example is not so minimal anymore, and reads less like a tutorial on "how to set up a domain-parallel training loop for a ViT" and more like a perf benchmark/evaluation suite. It's nice to two-birds and get useful performance plots out of it, but we could consider setting up more of an actual minimal/standalone demo train script in a separate example at some point (it could include demonstration of the automatic save/load checkpoint utilities, for example).

**Comment by coreyjadams** [comment]
@ktangsali you got pinged as CODEOWNER of the interrogate list-of-missing-docstrings.  But I exclusively pruned the list, so, should be all good :)
